generator/EnumGen.cs

```diff
@@ -25,49 +25,103 @@
 
 namespace GtkSharp.Generation {
 
-	public class EnumGen : GenBase {
+	// <enumeration name="BookmarkFileError" c:type="GBookmarkFileError" glib:error-domain="g-bookmark-file-error-quark">
+	//   <doc xml:space="preserve">Error codes returned by bookmark file parsing.</doc>
+	//   <member name = "invalid_uri" value="0" c:identifier="G_BOOKMARK_FILE_ERROR_INVALID_URI">
+	//     <doc xml:space="preserve">URI was ill-formed</doc>
+	//   </member>
+	public class EnumGen : GenBase
+	{
+		// TODO:
+		// See marshalling without casting
+
+		class EnumMember
+		{
+			public string Name { get; private set; }
+			public string Value { get; private set; }
+			bool isBitField;
+
+			public EnumMember (XmlElement elem, bool isBitField)
+			{
+				this.isBitField = isBitField;
+
+				Name = elem.GetAttribute (Constants.Name);
+				Value = elem.GetAttribute (Constants.Value);
+			}
+
+			public bool Validate ()
+			{
+				if (IsUnsigned) {
+					Value = Value.TrimEnd ('u', 'U');
+				}
+				if (isBitField) {
+					int value = int.Parse (Value);
+					Value = string.Format ("0x{0:X}", value);
+				}
+				return true;
+			}
 
-		string enum_type = String.Empty;
-		List<string> members = new List<string> ();
+			public bool IsUnsigned {
+				get {
+					return Value.EndsWith ("U", StringComparison.OrdinalIgnoreCase);
+				}
+			}
 
-		public EnumGen (XmlElement ns, XmlElement elem) : base (ns, elem)
+			public string Generate ()
+			{
+				return Name + " = " + Value + ",";
+			}
+		}
+
+		string enum_type = String.Empty;
+		readonly List<EnumMember> members = new List<EnumMember> ();
+		bool isBitField;
+
+		public EnumGen (XmlElement ns, XmlElement elem, bool isBitField) : base (ns, elem)
 		{
+			this.isBitField = isBitField;
 			foreach (XmlElement member in elem.ChildNodes) {
-				if (member.Name != "member")
-					continue;
-
-				string result = "\t\t" + member.GetAttribute(Constants.Name);
-				if (member.HasAttribute(Constants.Value)) {
-					string value = member.GetAttribute(Constants.Value);
-					if (value.EndsWith("U")) {
-						enum_type = " : uint";
-						value = value.TrimEnd('U');
-					}
-					result += " = " + value;
+				switch (member.Name) {
+				case Constants.Member:
+					var em = new EnumMember (member, isBitField);
+					members.Add (em);
+					break;
+				default:

```

Comment:
Enums/Bitfields may have methods in gir.

Reply:
Yeah, that's why I added the default case, to see what we're failing to generate.